src/cmd_line/commands/breakpoints.ts

```diff
@@ -0,0 +1,271 @@
+import * as vscode from 'vscode';
+import * as path from 'path';
+import { VimState } from '../../state/vimState';
+import { ExCommand } from '../../vimscript/exCommand';
+import { all, alt, eof, optWhitespace, regexp, seqObj, string, succeed, whitespace } from 'parsimmon';
+import { numberParser } from '../../vimscript/parserUtils';
+
+function isSourceBreakpoint(b: any): b is vscode.SourceBreakpoint {
+  return (b as vscode.SourceBreakpoint).location !== undefined;
+}
+function isFunctionBreakpoint(b: any): b is vscode.FunctionBreakpoint {
+  return (b as vscode.FunctionBreakpoint).functionName !== undefined;
+}
+
+/**
+ * Add Breakpoint Command
+ */
+type AddBreakpointExpr = { expr: string };
+type AddBreakpointFunction = { function: string };
+type AddBreakpointFile = { line: number; file: string };
+type AddBreakpointHere = 'here';
+type AddBreakpoint =
```

Comment:
I'd prefer a discriminated union here (with a `type` string enum). I think
```typescript
if (this.addBreakpoint.type === 'file') {...}
```
is cleaner than custom type guards in cases like this